server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -391,9 +391,21 @@ private static void innerParseObject(ParseContext context, ObjectMapper mapper,
             if (token == XContentParser.Token.FIELD_NAME) {
                 currentFieldName = parser.currentName();
                 paths = splitAndValidatePath(currentFieldName);
+
                 if (context.mapperService().isMetadataField(context.path().pathAsText(currentFieldName))) {
-                    throw new MapperParsingException("Field [" + currentFieldName + "] is a metadata field and cannot be added inside"
-                        + " a document. Use the index API request parameters.");
+                    if (context.mapperService().isFieldAllowedInSource(context.path().pathAsText(currentFieldName))) {
+                        // If token is a metadata field and is allowed in source, parse its value
```

Comment:
What's the reasoning for rejecting null and non-leaf values here?

Reply:
@jimczi and I had a discussion about the possible use cases we want to cover with this feature in the near future and agreed that for the shake of simplicity we should only accept non-null values (no objects or arrays). 

I don't have any strong opinions and I am happy to revisit this.

Reply:
I think null should be accepted but non-leaf values seemed more challenging. Now that I re-think about it, it shouldn't be an issue as long as we consume the value from the root (object or not).

Reply:
Ok, let me workon this again